src/coreclr/src/jit/codegencommon.cpp

```diff
@@ -6324,71 +6331,218 @@ void CodeGen::genZeroInitFrame(int untrLclHi, int untrLclLo, regNumber initReg,
+            blkSize = alignedLclHi - alignedLclLo;
+            assert((blkSize + alignmentLoBlkSize + alignmentHiBlkSize) == (untrLclHi - untrLclLo));
 
-#ifdef UNIX_AMD64_ABI
-        // Move back the argument registers
-        if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_RCX)
-        {
-            inst_RV_RV(INS_mov, REG_RCX, REG_R12);
-        }
+            if (untrLclLo != alignedLclLo)
+            {
+                assert(alignmentLoBlkSize > 0);
+                assert(alignmentLoBlkSize < XMM_REGSIZE_BYTES);
+                assert((alignedLclLo - alignmentLoBlkSize) == untrLclLo);
 
-        if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_RDI)
-        {
-            inst_RV_RV(INS_mov, REG_RDI, REG_R13);
-        }
-#else  // !UNIX_AMD64_ABI
-        // Move back the argument registers
-        if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_ECX)
-        {
-            inst_RV_RV(INS_mov, REG_ECX, REG_ESI);
-        }
-#endif // !UNIX_AMD64_ABI
+                zeroReg = genGetZeroReg(initReg, pInitRegZeroed);
 
-#else // TARGET*
+                int i = 0;
+                for (; i + REGSIZE_BYTES <= alignmentLoBlkSize; i += REGSIZE_BYTES)
+                {
+                    emit->emitIns_AR_R(ins_Store(TYP_I_IMPL), EA_PTRSIZE, zeroReg, frameReg, untrLclLo + i);
+                }
+                assert(i == alignmentLoBlkSize || (i + sizeof(int) == alignmentLoBlkSize));
+                if (i != alignmentLoBlkSize)
+                {
+                    emit->emitIns_AR_R(ins_Store(TYP_INT), EA_4BYTE, zeroReg, frameReg, untrLclLo + i);
+                    i += sizeof(int);
+                }
+
+                assert(i == alignmentLoBlkSize);
+            }
+#else // !defined(TARGET_AMD64)
+            // While we aren't aligning the start, we still want to
+            // zero anything that is not in a 16 byte chunk at end
+            int alignmentBlkSize   = blkSize & -XMM_REGSIZE_BYTES;
+            int alignmentHiBlkSize = blkSize - alignmentBlkSize;
+            int alignedLclHi       = untrLclLo + alignmentBlkSize;
+            blkSize                = alignmentBlkSize;
+
+            assert((blkSize + alignmentHiBlkSize) == (untrLclHi - untrLclLo));
+#endif
+            // The loop is unrolled 3 times so we do not move to the loop block until it
+            // will loop at least once so the threshold is 6.
+            if (blkSize < 6 * XMM_REGSIZE_BYTES)
+            {
+                // Generate the following code:
+                //
+                //   xorps   xmm4, xmm4
+                //   movups  xmmword ptr [ebp/esp-OFFS], xmm4
+                //   ...
+                //   movups  xmmword ptr [ebp/esp-OFFS], xmm4
+                //   mov      qword ptr [ebp/esp-OFFS], rax
+
+                emit->emitIns_R_R(INS_xorps, EA_ATTR(XMM_REGSIZE_BYTES), zeroSIMDReg, zeroSIMDReg);
+
+                int i = 0;
+                for (int regSize = XMM_REGSIZE_BYTES; i + regSize <= blkSize; i += regSize)
```

Comment:
Seems like this could simply be:
```suggestion
                int i = 0;
                for (; i < blkSize; i += XMM_REGSIZE_BYTES)
```